genienlp/data_utils/bootleg.py

```diff
@@ -0,0 +1,381 @@
+            if qid in ['Q6010', 'Q11401', 'Q511054', 'Q10460904', 'Q4955868']:
+                type = 'Q188451'
+
+            if title in ['song', 'single', 'musical composition', 'ballad', 'extended play', 'literary work',
+                         'television series', 'film', 'play']:
+                type = 'Q7366'
+            elif 'album' in title or title in []:
+                type = 'Q482994'
+            elif 'genre' in title or title in ['country', 'music by country or region', 'music term', 'republic',
+                                               'ethnic group', 'music scene', 'music style']:
+                type = 'Q188451'
+            elif 'person' in title or 'musician' in title or \
+                    title in ['singer', 'actor', 'musician', 'songwriter',
+                              'composer', 'singer-songwriter', 'musical group', 'drummer',
+                              'writer', 'poet', 'guitarist', 'rapper', 'painter',
+                              'film director', 'rock band', 'university teacher', 'journalist',
+                              'television presenter', 'saxophonist', 'music pedagogue',
+                              'association football player', 'disc jockey', 'record producer', 'engineer',
+                              'human biblical figure', 'big band', 'musical duo', 'girl group',
+                              'boy band', 'musical ensemble', 'artist', 'vocal group', 'heavy metal band',
+                              'literary character', 'lawyer', 'lyricist', 'baseball player']:
+                type = 'Q5'
+            
+            elif title in ['video game', 'disease', 'city of the United States', 'taxon',
+                           'Wikimedia disambiguation page', 'Wikimedia list article']:
+                type = 'unk'
+    
+    return type
+
+
+class Bootleg(object):
+    
+    def __init__(self, args):
+        self.args = args
+        
+        self.model_dir = f'{self.args.database_dir}/{self.args.bootleg_model}'
+        self.config_path = f'{self.model_dir}/bootleg_config.json'
+        self.cand_map = f'{self.args.database_dir}/wiki_entity_data/entity_mappings/alias2qids_wiki.json'
+
+        self.entity_dir = f'{self.args.database_dir}/wiki_entity_data'
+        self.embed_dir = f'{self.args.database_dir}/emb_data/'
+        
+        self.almond_domains = args.almond_domains
+        self.bootleg_post_process_types = args.bootleg_post_process_types
+        
+        with open(f'{self.args.database_dir}/emb_data/entityQID_to_wikidataTypeQID.json', 'r') as fin:
+            self.qid2type = ujson.load(fin)
+        with open(f'{self.args.database_dir}/es_material/type2id.json', 'r') as fin:
+            self.type2id = ujson.load(fin)
+
+        with open(f'{self.args.database_dir}/emb_data/wikidatatitle_to_typeid_0905.json', 'r') as fin:
+            title2typeid = ujson.load(fin)
+            self.typeid2title = {v: k for k, v in title2typeid.items()}
```

Comment:
What is the content of these three files? Are they just dictionaries?
If so, you can use the old [HashTable code](https://github.com/stanford-oval/genienlp/blob/69c0d5f801bf924a6a617edffcc6ca2c77971d66/genienlp/data_utils/hash_table.py) to create a binary on-disk representation of the dictionary that you can memory map. That will yield 0 memory usage, because the kernel will page in and out of disk as needed, and should solve our memory issues.

Reply:
Interesting! I'll look into it.